dropshots-gradle-plugin/src/main/kotlin/com/dropbox/dropshots/DropshotsPlugin.kt

```diff
@@ -1,122 +1,172 @@
-      srcDirs(referenceScreenshotDirectory)
-    }
+        // Create a test connected check task
+        addTasksForDeviceProvider(variant, deviceTestComponent, "connected", adbProvider)
 
-    val adbExecutablePath = provider { extension.adbExecutable.path }
-    extension.testVariants.all { variant ->
-      val testTaskProvider = variant.connectedInstrumentTestProvider
-      val variantSlug = variant.name.replaceFirstChar { if (it.isLowerCase()) it.titlecase(Locale.getDefault()) else it.toString() }
-
-      val screenshotDir = provider {
-        val appId = if (variant.testedVariant is ApkVariant) {
-          variant.testedVariant.applicationId
-        } else {
-          variant.packageApplicationProvider.get().applicationId
-          variant.applicationId
+        testedExtension.deviceProviders.forEach { deviceProvider ->
+          addTasksForDeviceProvider(variant, deviceTestComponent, deviceProvider.name, adbProvider)
         }
-        "/storage/emulated/0/Download/screenshots/$appId"
-      }
-
-      val clearScreenshotsTask = tasks.register(
-        "clear${variantSlug}Screenshots",
-        ClearScreenshotsTask::class.java,
-      ) {
-        it.adbExecutable.set(adbExecutablePath)
-        it.screenshotDir.set(screenshotDir)
-      }
-
-      val pullScreenshotsTask = tasks.register(
-        "pull${variantSlug}Screenshots",
-        PullScreenshotsTask::class.java,
-      ) {
-        it.adbExecutable.set(adbExecutablePath)
-        it.screenshotDir.set(screenshotDir.map { base -> "$base/diff" })
-        it.outputDirectory.set(testTaskProvider.flatMap { (it as AndroidTestTask).resultsDir })
-        it.finalizedBy(clearScreenshotsTask)
       }
+    }
+  }
 
-      val recordScreenshotsTask = tasks.register(
-        "record${variantSlug}Screenshots",
-        PullScreenshotsTask::class.java,
-      ) {
-        it.description = "Updates the local reference screenshots"
-
-        it.adbExecutable.set(adbExecutablePath)
-        it.screenshotDir.set(screenshotDir.map { base -> "$base/reference" })
-        it.outputDirectory.set(referenceScreenshotDirectory)
-        it.dependsOn(testTaskProvider)
-        it.finalizedBy(clearScreenshotsTask)
+  private fun Project.addTasksForDeviceProvider(
+    variant: Variant,
+    deviceTest: DeviceTest,
+    deviceProviderName: String,
+    adbProvider: Provider<File>,
+  ) {
+    tasks
+      .named {
+        it == "${deviceProviderName}${variant.name.capitalize()}${ComponentType.ANDROID_TEST_SUFFIX}"
       }
+      .all { testTask ->
+        val testDataProvider = when (testTask) {
+          is DeviceProviderInstrumentTestTask -> testTask.testData
+          is ManagedDeviceInstrumentationTestTask -> testTask.testData
+          is ManagedDeviceTestTask -> testTask.testData
+          else -> return@all
+        }
+        val defaultTestOutputDirectory = layout.buildDirectory
+          .dir("outputs/androidTest-results/$deviceProviderName/${variant.name}")
```

Comment:
I was having trouble figuring our `additionalTestOutputDir` and I don't think it's actually used because of the return in `testDataProvider` evaluation above.

I think collpasing `additionalTestOutputEnabled` and `additionalTestOutputDir` would be a lot more readable. 
```
val additionalTestOutputDir = when (testTask) {
    is DeviceProviderInstrumentTestTask -> testTask.additionalTestOutputEnabled.flatMap {
        if(it) { testTask.additionalTestOutputDir } else { provider { null } }
    }
    is ManagedDeviceInstrumentationTestTask -> testTask.getAdditionalTestOutputEnabled().flatMap {
        if(it) { testTask.getAdditionalTestOutputDir() } else { provider { null } }
    }
    is ManagedDeviceTestTask -> testTask.getAdditionalTestOutputEnabled().flatMap {
        if(it) { testTask.getAdditionalTestOutputDir() } else { provider { null } }
    }
    else -> return@configureEach
}
```

I would also consider combining it with the `testDataProvider` so that you only have a single `when` to read, but you'd need a data class of a pair return type so it's a little more subjective.